calicoctl/commands/convert.go

```diff
@@ -0,0 +1,148 @@
+  -f --filename=<FILENAME>      Filename to use to create the resource.  If set to
+                                "-" loads from stdin.
+  -o --output=<OUTPUT FORMAT>   Output format. One of: yaml or json.
+                                [Default: yaml]
+
+
+Description:
+  Convert config files between different API versions. Both YAML and JSON formats are accepted.
+
+  The default output will be printed to stdout in YAML format.
+`
+	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	if err != nil {
+		fmt.Printf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.\n", strings.Join(args, " "))
+		os.Exit(1)
+	}
+	if len(parsedArgs) == 0 {
+		return
+	}
+
+	var rp resourcePrinter
+	output := parsedArgs["--output"].(string)
+	// Only supported output formats are yaml (default) and json.
+	switch output {
+	case "yaml", "yml":
+		rp = resourcePrinterYAML{}
+	case "json":
+		rp = resourcePrinterJSON{}
+	default:
+		rp = nil
+	}
+
+	if rp == nil {
+		fmt.Printf("unrecognized output format '%s'\n", output)
+		os.Exit(1)
+	}
+
+	filename := argutils.ArgStringOrBlank(parsedArgs, "--filename")
+
+	// Load the V1 resource from file and convert to a slice
+	// of resources for easier handling.
+	resV1, err := v1resourceloader.CreateResourcesFromFile(filename)
+	if err != nil {
+		fmt.Printf("Failed to execute command: %v\n", err)
+		os.Exit(1)
+	}
+
+	var results []runtime.Object
+	for _, v1Resource := range resV1 {
+		v3Resource, err := convertResource(v1Resource)
+		if err != nil {
+			fmt.Printf("Failed to execute command: %v\n", err)
+			os.Exit(1)
+		}
+		results = append(results, v3Resource)
+	}
+
+	log.Infof("results: %+v", results)
+
+	err = rp.print(nil, results)
+	if err != nil {
+		fmt.Println(err)
+	}
+}
+
+func convertResource(v1resource unversioned.Resource) (runtime.Object, error) {
+	switch strings.ToLower(v1resource.GetTypeMetadata().Kind) {
+	case "node":
+		return convert(conversion.Node{}, v1resource)
+	case "hostendpoint":
+		return convert(conversion.HostEndpoint{}, v1resource)
+	case "workloadendpoint":
+		return convert(conversion.WorkloadEndpoint{}, v1resource)
+	case "profile":
+		return convert(conversion.Profile{}, v1resource)
+	case "policy":
+		return convert(conversion.Policy{}, v1resource)
+	case "ippool":
+		return convert(conversion.IPPool{}, v1resource)
+	case "bgppeer":
+		return convert(conversion.BGPPeer{}, v1resource)
+
+	default:
+		return nil, fmt.Errorf("conversion for the resource type '%s' is not supported", v1resource.GetTypeMetadata().Kind)
+	}
+}
+
+func convert(convRes conversion.Converter, v1resource unversioned.Resource) (conversion.Resource, error) {
+	// Convert v1 API resource to v1 backend KVPair.
+	kvp, err := convRes.APIV1ToBackendV1(v1resource)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert v1 backend KVPair to v3 API resource.
+	res, err := convRes.BackendV1ToAPIV3(kvp)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
We should probably validate the final result resource too.

Reply:
I think it would be a good idea, that would catch some of the problems I've seen in testing when taking converted data and applying it.

Reply:
hmmm, not sure if it makes sense to fail validation on something that user can't control, I'd say it's better to fail when a user tries to apply the config. Failing validation on the config we generated seems a bit weird (from the UX PoV)

Reply:
Agree with @gunjan5 on this - our bad v3 manifest will be caught when the user applies it. As G5 mentions, there's nothing the user can do about our botched conversion (other than edit and try to apply again).